controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -480,26 +480,22 @@ public String getScopeName() {
     }
 
     @Override
-    public CompletableFuture<VersionedTransactionData> pingTransaction(final UUID txId, final long lease) {
-        return getTransactionEpoch(txId).thenCompose(epoch -> getActiveTx(epoch, txId)
-                .thenCompose(data -> {
-                    ActiveTxnRecord activeTxnRecord = ActiveTxnRecord.parse(data.getData());
-                    if (activeTxnRecord.getTxnStatus() == TxnStatus.OPEN) {
-                        // Update txn record with new lease value and return versioned tx data.
-                        ActiveTxnRecord newData = new ActiveTxnRecord(activeTxnRecord.getTxCreationTimestamp(),
-                                System.currentTimeMillis() + lease, activeTxnRecord.getMaxExecutionExpiryTime(),
-                                activeTxnRecord.getScaleGracePeriod(), activeTxnRecord.getTxnStatus());
-
-                        return updateActiveTx(epoch, txId, newData.toByteArray())
-                                .thenApply(x ->
-                                        new VersionedTransactionData(epoch, txId, data.getVersion() + 1,
-                                                TxnStatus.OPEN, activeTxnRecord.getTxCreationTimestamp(),
-                                                activeTxnRecord.getMaxExecutionExpiryTime(),
-                                                activeTxnRecord.getScaleGracePeriod()));
-                    } else {
-                        return FutureHelpers.failedFuture(new IllegalStateException(txId.toString()));
-                    }
-                }));
+    public CompletableFuture<VersionedTransactionData> pingTransaction(final VersionedTransactionData txnData,
+                                                                       final long lease) {
+        // Update txn record with new lease value and return versioned tx data.
+        final int epoch = txnData.getEpoch();
+        final UUID txnId = txnData.getId();
+        final int version = txnData.getVersion();
+        final long creationTime = txnData.getCreationTime();
+        final long maxExecutionExpiryTime = txnData.getMaxExecutionExpiryTime();
+        final long scaleGracePeriod = txnData.getScaleGracePeriod();
+        final TxnStatus status = txnData.getStatus();
+        final ActiveTxnRecord newData = new ActiveTxnRecord(creationTime, System.currentTimeMillis() + lease,
+                maxExecutionExpiryTime, scaleGracePeriod, status);
+        final Data<Integer> data = new Data<>(newData.toByteArray(), version);
+
+        return updateActiveTx(epoch, txnId, data).thenApply(x -> new VersionedTransactionData(epoch, txnId,
+                version + 1, status, creationTime, maxExecutionExpiryTime, scaleGracePeriod));
```

Comment:
since this version is returned by the underlying store, it should be opaque here.. we should not do this math and add 1..